core/sdk.go

```diff
@@ -35,66 +37,188 @@ type PolicyResult struct {
 // Warning: This interface is experimental, and it could change with breaking also in rond patches.
 // Does not use outside this repository until it is not ready.
 type SDK interface {
-	// Warning: this method will be removed in the near future. Do not use it outside Rond.
-	Metrics() metrics.Metrics
-
 	FindEvaluator(logger *logrus.Entry, method, path string) (SDKEvaluator, error)
-	EvaluatorFromConfig(logger *logrus.Entry, config openapi.RondConfig) SDKEvaluator
 }
 
 // Warning: This interface is experimental, and it could change with breaking also in rond patches.
 // Do not use outside this repository until it is not ready.
 type SDKEvaluator interface {
 	Config() openapi.RondConfig
-	PartialResultsEvaluators() PartialResultsEvaluators
+
+	EvaluateRequestPolicy(ctx context.Context, req RondInput, userInfo types.User) (PolicyResult, error)
+	EvaluateResponsePolicy(ctx context.Context, rondInput RondInput, userInfo types.User, decodedBody any) ([]byte, error)
 }
 
 type evaluator struct {
 	rond       rondImpl
 	logger     *logrus.Entry
 	rondConfig openapi.RondConfig
+
+	routeInfo openapi.RouterInfo
+}
+
+func (e evaluator) metrics() metrics.Metrics {
+	return e.rond.metrics
+}
+
+func (e evaluator) partialResultEvaluators() PartialResultsEvaluators {
+	return e.rond.partialResultEvaluators
 }
 
 func (e evaluator) Config() openapi.RondConfig {
 	return e.rondConfig
 }
 
-func (e evaluator) PartialResultsEvaluators() PartialResultsEvaluators {
-	return e.rond.evaluator
+func (e evaluator) EvaluateRequestPolicy(ctx context.Context, req RondInput, userInfo types.User) (PolicyResult, error) {
+	if req == nil {
+		return PolicyResult{}, fmt.Errorf("RondInput cannot be empty")
+	}
+
+	rondConfig := e.Config()
+
+	input, err := req.FromRequestInfo(userInfo, nil)
+	if err != nil {
+		return PolicyResult{}, err
+	}
+
+	regoInput, err := CreateRegoQueryInput(e.logger, input, RegoInputOptions{
+		EnableResourcePermissionsMapOptimization: rondConfig.Options.EnableResourcePermissionsMapOptimization,
+	})
+	if err != nil {
+		return PolicyResult{}, nil
+	}
+
+	var evaluatorAllowPolicy *OPAEvaluator
+	if !rondConfig.RequestFlow.GenerateQuery {
+		evaluatorAllowPolicy, err = e.partialResultEvaluators().GetEvaluatorFromPolicy(ctx, rondConfig.RequestFlow.PolicyName, regoInput, e.rond.evaluatorOptions)
+		if err != nil {
+			return PolicyResult{}, err
+		}
+	} else {
+		evaluatorAllowPolicy, err = e.rond.opaModuleConfig.CreateQueryEvaluator(ctx, e.logger, rondConfig.RequestFlow.PolicyName, regoInput, e.rond.evaluatorOptions)
+		if err != nil {
+			return PolicyResult{}, err
+		}
+	}
+
+	opaEvaluationTimeStart := time.Now()
```

Comment:
I'd expect the core to be responsible for metrics measurement; we are adding loads of statement in the metrics computation which I'd not expect since what I need from these metrics is the timing of my policies without the overhead produced by rönd 😞

Reply:
>The overhead is equal to all the response, so it does not changes from policy to policy (which is the really important thing) when we compare something.

I agree on this, there would still be a change with previous metrics we must be aware of when upgrading the service in our application though

>The overhead is really low since there isn't heavy logic inside it.

Did you measure it? I mean: I agree on practice however there is still an unmeasured overhead and we don't know the impact it would have on the metrics 🤔 

>From my point of view, metrics are more sdk related since them are based on the technology used to retrieve it (in our case, prometheus), and we could add more in the future.

I agree on this, especially on the dependencies matter; what I was expecting was some kind of interface that the core would use to measure performances and provide directly to who is implementing rönd; using prometheus or something else should be up to the final user (the sidecar service would implement prometheus, while a third party may implement their own metrics interface to map on their own metrics system). In this sense I'd remove the dependency of the metrics system from the sdk as well but keeping measurement as close as possible to the actual evaluation.. How do you see this? If you agree on this I'd revert changes to move back the prom monitoring in the core and then extract it as interface in another PR

Reply:
> Did you measure it? I mean: I agree on practice however there is still an unmeasured overhead and we don't know the impact it would have on the metrics 🤔

I have not measured, but there is an immediate return or a type assertion + len + return of the data. Nothing which has high computational complexity.

> I agree on this, especially on the dependencies matter; what I was expecting was some kind of interface that the core would use to measure performances and provide directly to who is implementing rönd; using prometheus or something else should be up to the final user (the sidecar service would implement prometheus, while a third party may implement their own metrics interface to map on their own metrics system). In this sense I'd remove the dependency of the metrics system from the sdk as well but keeping measurement as close as possible to the actual evaluation.. How do you see this? If you agree on this I'd revert changes to move back the prom monitoring in the core and then extract it as interface in another PR

Yes, it is possible. But, as I say, I'm not sure it is ok that the core add metrics and logging, as those are more dependent on the usage of the feature.
Moreover, we should also support the management of nil metrics interface (which sdk already support, since it init the metrics itself, but it is not possible in the core). Because I think it's correct to set the metrics as required params. 

I would add a new option parameter which can contains the metrics and logging, and if you pass it it will be used (or something similar)

Reply:
>I have not measured, but there is an immediate return or a type assertion + len + return of the data. Nothing which has high computational complexity.

I'm more worried on what may be added in the future and won't take into account the eval is under performance monitoring; shall we document the function to say this at least?

>Because I think it's correct to set the metrics as required params.

Did you mean in the SDK or in the core?